File: solutions/elixir/bob/1/lib/bob.ex
defmodule Bob do
  @spec hey(String.t()) :: String.t()
  def hey(input) do
    cond do
      String.trim(input) == "" -> "Fine. Be that way!"
      yelling?(input) and questioning?(input) -> "Calm down, I know what I'm doing!"
      questioning?(input) -> "Sure."
      yelling?(input) -> "Whoa, chill out!"
      true -> "Whatever."
    end
  end

  # upcase result is the same, and there's at least 1 uppercase letter in input
  defp yelling?(input), do: String.upcase(input) == input and has_upper?(input)
  defp questioning?(input) do
    input = String.trim(input)
    String.at(input, String.length(input)-1) == "?"
  end

  defp has_upper?(input) do
    input =~ ~r/.*[^[:digit:][:punct:][:blank:]]/
  end
end

File: solutions/haskell/bob/1/src/Bob.hs
module Bob (responseFor) where

import Data.Text (Text, strip, toUpper, empty, last, pack, filter)
import Data.Char (isAlpha)

isSilence :: Text -> Bool
isSilence  = (empty ==)

isQuestion :: Text -> Bool
isQuestion t = not (isSilence t) && Data.Text.last t == '?'

isYelling :: Text -> Bool
isYelling t = not (isSilence t) && t == toUpper t && Data.Text.filter isAlpha t /= empty

responseFor :: Text -> Text
responseFor xs
  | isSilence $ strip xs = pack "Fine. Be that way!"
  | let stripped = strip xs in isQuestion stripped && isYelling stripped = pack "Calm down, I know what I'm doing!"
  | isQuestion $ strip xs = pack "Sure."
  | isYelling $ strip xs = pack "Whoa, chill out!"
  | otherwise = pack "Whatever."
  
